core/vm/ewasm.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of go-ethereum.
+//
+// go-ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// go-ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+
+// This file lists the EEI functions, so that they can be bound to any
+// eWASM-compatible module, as well as the types of these functions
+
+package vm
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/go-interpreter/wagon/wasm"
+
+	"github.com/go-interpreter/wagon/exec"
+)
+
+// InterpreterEWASM implements the Interpreter interface for eWASM.
+type InterpreterEWASM struct {
+	vm *exec.VM
+
+	evm *EVM
+
+	StateDB StateDB
+
+	gasTable params.GasTable
+
+	readOnly bool
+
+	contract *Contract
+
+	returnData []byte
+}
+
+// NewEWASMInterpreter creates a new wagon-based eWASM interpreter. It
+// currently takes a *vm.EVM pointer as a proxy to the client's internal
+// state; this will be fixed in subsequent updates.
+func NewEWASMInterpreter(evm *EVM, cfg Config) Interpreter {
+	return &InterpreterEWASM{StateDB: evm.StateDB, evm: evm, gasTable: evm.chainConfig.GasTable(evm.BlockNumber)}
+}
+
+// Run loops and evaluates the contract's code with the given input data and returns
+// the return byte-slice and an error if one occurred.
+func (in *InterpreterEWASM) Run(contract *Contract, input []byte) ([]byte, error) {
+	in.contract = contract
+	in.contract.Input = input
+
+	module, err := wasm.ReadModule(bytes.NewReader(contract.Code), WrappedModuleResolver(in))
+	if err != nil {
+		return nil, fmt.Errorf("Error decoding module at address %s: %v", contract.Address(), err)
+	}
+	vm, err := exec.NewVM(module)
+	if err != nil {
+		return nil, fmt.Errorf("could not create the vm: %v", err)
+	}
+	vm.RecoverPanic = true
+	in.vm = vm
+
+	var ret []byte
+
+	// If the module doesn't have any start function, check if the
+	// exports specify a `main` function and if so, execute it.
+	if module.Start == nil {
```

Comment:
So is `exec.NewVM` executing the start function?

Since this is disabled in ewasm, you could just leave out the if (perhaps replace it with a panic).

In any case you need to add support for executing system contracts, at least in the create case, so that incoming bytecode can be run through metering.